Vector.go, matrix.go and mgl64 generation

alexvasi opened the pull request:
Fix for issues #34 and #32

Run `go generate mgl32` to make `vector.go`, `matrix.go` and `mgl64` (or simply `go generate` in mgl32 directory). I would appreciate any ideas or suggestions.

All changes to the lib are purely cosmetic (comments, formatting) apart from following:
* `matstack/transformStack` is added to `mgl64`.
* mgl64/util.go:`MinValue = math.SmallestNonzeroFloat64`.
* Left alone type casts like `float64(math.Sqrt(float64(...)))`. I do believe it's a noop.
* Additional tests for vectors.
* Moved vectorStatic and matrixStatic files to templates.
* Order of vector and matrix functions is changed. Diffs are messy. To get nice diffs one can make a couple of trivial changes to old `genprog` (I will gladly send it to anyone on request).

shurcooL commented on mgl32/codegen.go:
```diff
@@ -0,0 +1,273 @@
+	flag.Parse()
+	if flag.NArg() > 0 || flag.NFlag() == 0 {
+		flag.Usage()
+		os.Exit(1)
```
Minor, but this should be `os.Exit(2)`. Exit code 2 is typically used when flags are not as expected.

Jragonmiris replied:
Seems good to me, but I'm dubious about moving the stack into the main
package. We were working on having more subpackages rather than fewer.

On Sat, Mar 14, 2015 at 10:06 AM Dmitri Shuralyov <notifications@github.com>
wrote:

> In mgl32/codegen.go
> <https://github.com/go-gl/mathgl/pull/39#discussion_r26439497>:
>
> > +func main() {
> > +	flag.Usage = func() {
> > +		fmt.Println("Usage: codegen -template file.tpl -output file.go")
> > +		fmt.Println("Usage: codegen -mgl64 [-dir ../mgl64]")
> > +		flag.PrintDefaults()
> > +	}
> > +
> > +	tplPath := flag.String("template", "file.tpl", "template path")
> > +	oPath := flag.String("output", "file.go", "output path")
> > +	mgl64 := flag.Bool("mgl64", false, "make mgl64")
> > +	mgl64Path := flag.String("dir", "../mgl64", "path to mgl64 location")
> > +
> > +	flag.Parse()
> > +	if flag.NArg() > 0 || flag.NFlag() == 0 {
> > +		flag.Usage()
> > +		os.Exit(1)
>
> Minor, but this should be os.Exit(2). Exit code 2 is typically used when
> flags are not as expected.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/go-gl/mathgl/pull/39/files#r26439497>.
>

shurcooL commented:
Really good work, thank you for the PR! I will review soon.

One point, in previous issue you said:

> First of all it works. It is a zero-diff change... well, not actually. The order of functions in vector.go is different now. It takes 10 or so additional lines in template to make the order the same. I dumped them. But that's how I tested new generator.

Would you mind including those 10 lines (if you already have them anyway) in this PR to make the generated output diff smaller, making it easier to review. Then we can remove those 10 lines in a follow up PR.

shurcooL commented:
> Run `go generate mgl32` to make `vector.go`, `matrix.go` and `mgl64` (or simply `go generate` in mgl32 directory). I would appreciate any ideas or suggestions.

I tested it out, and it works okay. Note that you need to specify import path for `go generate`, so it should either be `go generate github.com/go-gl/mathgl/mgl32` or using relative import paths, `go generate ./mgl32`. Doing simply `go generate mgl32` will fail because `mgl32` is not a correct import path.

If you're looking for suggestions, I think it would be better if each package were more standalone and only generated itself. So,

```
# This generates the mgl32 package only.
go generate github.com/go-gl/mathgl/mgl32

# This generates the mgl64 package only.
go generate github.com/go-gl/mathgl/mgl64
```

And of course it'd be possible to generate all mathgl packages by doing `go generate ./...` while inside the repo root. This is just a suggestion/optional improvement.

I tested it out and everything seems fine, I am not seeing any issues. Please address my `os.Exit(2)` comment and it has my LGTM.

shurcooL commented:
I have one more minor suggestion. If I'm not mistaken, it's more common to use ".tmpl" as the file extension for Go html or text/template files. I think it'd be more readable and consistent compared to ".tpl".

alexvasi commented:
Sure. Fixed README, os.exit and template extensions. 527fa91328ba5337e36ce896c0c6636bca3c556c

> I think it would be better if each package were more standalone and only generated itself.

I know, it is somewhat counterintuitive to execute `go generate ./mgl32` to make `mgl64`. And it's definitely a thing that could be improved. Mgl is 2 separate packages only for client code. I see all this code generation as a way of writing float32/64 generic code. Is there really a need for another command to generate 64-bit version? I thought that `mgl64` have to mirror `mgl32` and tried to enforce it by workflow. Maybe, it would be nice to add some checks to travis later.

> You could also move the codegen.go file outside of mgl32 folder

I see your point, it belongs outside of mgl32 logically. I tried to find some notable examples of `generate` usage (http://golang.org/src/encoding/gob/, http://golang.org/src/time/) and it looks like including gen file with `+build ignore` flag is the way to go, no pun intended. At the same time I'm not entirely sure what's the best approach here and completely missed the way how it's done in go-gl/glew for example. I do want to make this code as consistent with other go-gl projects as possible. So I tried to move codegen.go and template files to root directory and still keep `go generate` directives symmetric... the outcome felt too "hackish" to commit. After that I moved `go generate` lines from util.go to separate files in each directory and experienced the urge to go back to Makefile. Yeah, I know) It's more or less organizational changes in the end and we could do it later.

shurcooL commented:
@Jragonmiris, since this is largely your codebase, any thoughts on this? How do you feel about the template files and their readability? Does this look like a positive change for future maintainability of the library?

Jragonmiris commented:
Sorry, I won't be able to look at this until Sunday, but I will then.

On Tue, Mar 17, 2015 at 9:53 PM Dmitri Shuralyov <notifications@github.com>
wrote:

> @Jragonmiris <https://github.com/Jragonmiris>, since this is largely your
> codebase, any thoughts on this? How do you feel about the template files
> and their readability? Does this look like a positive change for future
> maintainability of the library?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/go-gl/mathgl/pull/39#issuecomment-82738835>.
>

shurcooL commented:
Thanks @Jragonmiris.

Also, please take a look at https://github.com/google/gxui/issues/28#issuecomment-83750220 - they seem to be considering using `mathgl`, I suggested that contributions to make `mathgl` better as they see fit would be likely very welcome. Perhaps you'd want to comment on that.

Jragonmiris commented:
I still haven't looked at it and won't until later today, however, I'd like to point out that since we internally moved to the new img/f{32/64} types, as I recall the specification said that those types are RMO by definition. (I recall because I was the one who suggested it).

This would be a huge breaking change, but new codegen files may be the time to do it.

Jragonmiris commented:
Seems fine to me, other than what I just brought up.

shurcooL commented:
Can you please elaborate on what would be a breaking change?

I don't think this PR should introduce any breaking changes, it's goal should be to generate identical code, the only difference is in code organization. If needed, breaking changes can be done later in another PR (with appropriate discussion).

shurcooL commented:
> Seems fine to me,

Do you think it's an improvement over the previous approach? Will it be better in the long run?

Jragonmiris commented:
Internally we switched to the image/f32 and image/f64 types, correct? We defined those to be RMO, but mathgl is CMO. A new template would probably be the best time to switch between RMO and CMO, though the change would break pretty much everything.

Otherwise, this seems like a big improvement in maintenance.

Only other thing is that @alexvasi should add themselves to authors/contributors.

shurcooL commented:
> Otherwise, this seems like a big improvement in maintenance.

Great, that's the most important thing for this PR.

As I said, I don't think this PR should do any API or functionality changes. Thought it make it easier in the future.

Yes, we use `x/image/math/f32` types internally (only for Vec types and Mat3, Mat4), but that's an implementation detail. The public types of `mathgl` remain as before, their definition does not change.

If you do want to change between RMO/CMO, that is definitely material for a separate issue/PR/discussion.

shurcooL commented:
@Jragonmiris, any updates? I'm still not completely sure if this has your LGTM or not.

Should we merge this, or is there something else that needs to be done?

shurcooL commented:
Friendly ping.

Jragonmiris commented:
Actually, I missed something. @alexvasi, please submit another PR to add yourself to AUTHOR and CONTRIBUTORS, I'll merge it right away. :)
